### PR TITLE
vreportf: Fix interleaving issues, remove 4096 limitation

### PR DESCRIPTION
**Pull request by SyntevoAlex**
This fixes t5516 on Windows.
For detailed explanation please refer to code comment in vreportf().

There was a lot of back-and-forth already in vreportf():
d048a96e (2007-11-09) - 'char msg[256]' is introduced to avoid interleaving
389d1767 (2009-03-25) - Buffer size increased to 1024 to avoid truncation
625a860c (2009-11-22) - Buffer size increased to 4096 to avoid truncation
f4c3edc0 (2015-08-11) - Buffer removed to avoid truncation
b5a9e435 (2017-01-11) - Reverts f4c3edc0 to be able to replace control
                        chars before sending to stderr

This fix attempts to solve all issues:
1) avoid multiple fprintf() interleaving
2) avoid truncation
3) avoid char interleaving in fprintf() on some platforms
4) avoid buffer block interleaving when output is large
5) avoid Out-of-order messages
6) replace control characters in output

Other commits worthy of notice:
9ac13ec9 (2006-10-11) - Another attempt to solve interleaving.
						This is seemingly related to d048a96e.
137a0d0e (2007-11-19) - Addresses out-of-order for display()
34df8aba (2009-03-10) - Switches xwrite() to fprintf() in recv_sideband()
						to support UTF-8 emulation
eac14f89 (2012-01-14) - Removes the need for fprintf() for UTF-8 emulation,
						so it's safe to use xwrite() again
5e5be9e2 (2016-06-28) - recv_sideband() uses xwrite() again

Signed-off-by: Alexandr Miloslavskiy <alexandr.miloslavskiy@syntevo.com>

### CODE REVIEW

**Comment by dscho** [comment]
Please ignore the osx failures for now... working on it.

**Comment by dscho** [comment]
FWIW I think that I was able to present a much less controversial fix in #428.

File: usage.c [hunk 1]
```diff
@@ -6,17 +6,40 @@
...
-	fprintf(stderr, "%s%s\n", prefix, msg);
+	prefix_size = strlen(prefix);
+	if (prefix_size >= sizeof(msg))
+		BUG("vreportf: prefix is too long");
+
+	memcpy(msg, prefix, prefix_size);
+
+	ret = vsnprintf(msg + prefix_size, sizeof(msg) - prefix_size, err, params);
+	if (ret < 0)
+		BUG("your vsnprintf is broken (returned %d)", ret);
```
Comment (dscho) [inline usage.c]: That'll cause an infinite loop, no?

Comment (SyntevoAlex) [reply usage.c]: Not to my knowledge, because BUG() uses a different format string, which isn't broken.

Comment (SyntevoAlex) [reply usage.c]: I'm currently testing this patch. Any feedback welcome, seriously. I do try to achieve high quality, and it takes some time.

**Comment by SyntevoAlex** [comment]
I'm currently testing this patch. Any feedback welcome, seriously. I do try to achieve high quality, and it takes some time.

File: usage.c (hunk 1 above)
Comment (SyntevoAlex) [reply usage.c]: Do you agree?

**Comment by SyntevoAlex** [comment]
Looks good to me now. Any objections?

**Comment by SyntevoAlex** [comment]
All of sudden, a failure on Linux in `t5504-fetch-receive-strict.sh` test 8.
Another broken pipe, I imagine.

**Comment by dscho** [comment]
> All of sudden, a failure on Linux in `t5504-fetch-receive-strict.sh` test 8.
> Another broken pipe, I imagine.

It's another known-flaky test... See https://github.com/gitgitgadget/git/issues/258 for details.

File: usage.c (hunk 1 above)
Comment (dscho) [reply usage.c]: In my opinion, this is too fragile. If we're here, things are already going wrong, and I'd rather not add a recursion into the same function.

**Comment by SyntevoAlex** [comment]
> It's another known-flaky test... See #258 for details.

:(

File: usage.c (hunk 1 above)
Comment (SyntevoAlex) [reply usage.c]: Just exit silently?

Comment (SyntevoAlex) [reply usage.c]: I would say, if `BUG()` is trustworthy, then it shouldn't be a problem to have 1 level of recursion. If `BUG()` is suspected to do very weird things, then it should rather be fixed.

File: usage.c [hunk 2]
```diff
@@ -6,17 +6,40 @@
...
+
+	total_size = strlen(msg);   /* vsnprintf() returns _desired_ size */
+	msg[total_size++] = '\n';   /* it's ok to overwrite terminating NULL */
```
Comment (dscho) [inline usage.c]: It's a `NUL`, not a `NULL`.

Comment (SyntevoAlex) [reply usage.c]: https://linux.die.net/man/3/fprintf
> ... the terminating null ...

File: usage.c (hunk 1 above)
Comment (dscho) [reply usage.c]: `BUG()` calls `vreportf()`. This recursion makes me uneasy. You might say that it cannot cause problems. But even you will have to admit that reasoning about this is not easy, and therein lies the opposite of perfect.

In any case, Git requires a `vsnprintf()` that returns the desired size, otherwise pretty much every `strbuf_addf()` call is broken (and there are many). So yes, if things go _so_ wrong as to let `vsnprintf()` return something negative, yes, I would exit silently.

And I would _certainly_ not introduce such a check in a commit that does many other things at the same time, too.

File: usage.c (hunk 2 above)
Comment (SyntevoAlex) [reply usage.c]: copied from there.

File: usage.c [hunk 3]
```diff
@@ -6,17 +6,40 @@
...
+
+	replace_control_chars(msg, total_size);
```
Comment (dscho) [inline usage.c]: I guess I find it totally unnecessary to refactor this part of the code into its own function. It's not like we introduce another caller, or that those four lines were particularly complex.

In any case, a refactoring needs to go into its own commit.

File: usage.c [hunk 4]
```diff
@@ -6,17 +6,40 @@
...
+
+	fflush(stderr);             /* flush FILE* before writing lowlevel fd */
```
Comment (dscho) [inline usage.c]: This comment just repeats what the statement says, and is therefore non-DRY.

A more interesting bit of information is *why* this is necessary, and this information naturally should go into the commit message, not into the code, as code comments like ones talking about specifics of a certain platform are _prone_ to become stale, which we try to avoid in Git's source code. The reasoning behind a patch is pretty much tied to the commit itself, and should therefore be described in the commit message.

Speaking of commit messages: the style you chose deviates noticeably from the style found in the existing commit history, and I wonder what reasons there are for that.

Comment (SyntevoAlex) [reply usage.c]: Can you suggest a better wording? It's necessary because FILE* and fd are different layers of abstraction. I think that problems can occur on any platform.

File: usage.c [hunk 5]
```diff
@@ -6,17 +6,40 @@
...
+	xwrite(2, msg, total_size); /* writing directly to fd is most atomic */
```
Comment (dscho) [inline usage.c]: Having seen this function, I see this PR edging closer and closer, step by step, to what I proposed already. There are still a couple of unnecessary bits and pieces that could be chopped off to arrive there, but I guess I really don't see what big improvements this PR offers compared to my proposed code (which was very similar to the code Jeff King proposed originally).

But maybe I am still missing something crucial?

File: usage.c (hunk 4 above)
Comment (SyntevoAlex) [reply usage.c]: As for commit message, what do you mean by style? Headers? I thought there's too much information to put in one stream, easier to navigate with headers. Just like in any other text.

File: usage.c (hunk 5 above)
Comment (SyntevoAlex) [reply usage.c]: I think the absence of bugs counts much more then small personal preferences about this and that.

File: usage.c (hunk 4 above)
Comment (dscho) [reply usage.c]: No, I cannot suggest a better wording, as my suggestion to remove the unnecessary comment still stands.

File: usage.c (hunk 2 above)
Comment (dscho) [reply usage.c]: Okay, I don't think you want to hear what I have to say, so I'll just stop.

File: usage.c (hunk 1 above)
Comment (SyntevoAlex) [reply usage.c]: I will change that to exiting silently.

File: usage.c (hunk 3 above)
Comment (SyntevoAlex) [reply usage.c]: Looked like a good function candidate to me. I guess git coding style also suggests to give names instead of writing comments. As for commit granularity, I'm not used to it. OK, I will inline the function again.

File: usage.c (hunk 4 above)
Comment (SyntevoAlex) [reply usage.c]: I don't expect that an average reader will understand quickly why fflush() could be needed here, so sounds like exactly when comments are good?

How about this wording? `We're going to write to low-level fd while higher-level FILE* misght still have something cached`
